File: android/designsystem/src/main/kotlin/com/bolsillo/designsystem/theme/BolsilloTheme.kt
package com.bolsillo.designsystem.theme

import androidx.compose.foundation.isSystemInDarkTheme
import androidx.compose.material3.LocalContentColor
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.ProvideTextStyle
import androidx.compose.runtime.Composable
import androidx.compose.runtime.CompositionLocalProvider
import androidx.compose.runtime.ReadOnlyComposable

/**
 * Bolsillo's Compose theme. Provides Material3 [androidx.compose.material3.MaterialTheme] for
 * standard slots and BolsilloTheme.* CompositionLocals for app-specific roles
 * (extra colors, typography, shapes, spacing, elevation) that M3 doesn't model.
 *
 * Reads light/dark token sets per [isSystemInDarkTheme]; do not hardcode colors
 * or sizes in composables (Article VI — visual parity).
 */
@Composable
fun BolsilloTheme(
    darkTheme: Boolean = isSystemInDarkTheme(),
    content: @Composable () -> Unit,
) {
    val colors = if (darkTheme) DarkBolsilloColors else LightBolsilloColors
    val typography = DefaultBolsilloTypography
    val shapes = DefaultBolsilloShapes

    CompositionLocalProvider(
        LocalBolsilloColors provides colors,
        LocalBolsilloTypography provides typography,
        LocalBolsilloShapes provides shapes,
        LocalBolsilloSpacing provides DefaultBolsilloSpacing,
        LocalBolsilloElevation provides DefaultBolsilloElevation,
        LocalContentColor provides colors.textPrimary,
    ) {
        MaterialTheme(
            colorScheme = colors.toColorScheme(),
            typography = typography.toMaterialTypography(),
            shapes = shapes.toMaterialShapes(),
        ) {
            ProvideTextStyle(value = typography.body, content = content)
        }
    }
}

object BolsilloTheme {
    val colors: BolsilloColors
        @Composable @ReadOnlyComposable
        get() = LocalBolsilloColors.current
    val typography: BolsilloTypography
        @Composable @ReadOnlyComposable
        get() = LocalBolsilloTypography.current
    val shapes: BolsilloShapes
        @Composable @ReadOnlyComposable
        get() = LocalBolsilloShapes.current
    val spacing: BolsilloSpacing
        @Composable @ReadOnlyComposable
        get() = LocalBolsilloSpacing.current
    val elevation: BolsilloElevation
        @Composable @ReadOnlyComposable
        get() = LocalBolsilloElevation.current
}

File: ios/Packages/Bolsillo/Sources/BolsilloDesignSystem/Theme/BolsilloTheme.swift
import SwiftUI

public struct BolsilloTheme {
    public let colors: BolsilloColors
    public let typography: BolsilloTypography
    public let spacing: BolsilloSpacing
    public let radius: BolsilloRadius
    public let elevation: BolsilloElevation
    public let gradients: BolsilloGradients

    public init(
        colors: BolsilloColors,
        typography: BolsilloTypography,
        spacing: BolsilloSpacing,
        radius: BolsilloRadius,
        elevation: BolsilloElevation,
        gradients: BolsilloGradients
    ) {
        self.colors     = colors
        self.typography = typography
        self.spacing    = spacing
        self.radius     = radius
        self.elevation  = elevation
        self.gradients  = gradients
    }

    public static let light = BolsilloTheme(
        colors:     .light,
        typography: .shared,
        spacing:    .shared,
        radius:     .shared,
        elevation:  .light,
        gradients:  .shared
    )

    public static let dark = BolsilloTheme(
        colors:     .dark,
        typography: .shared,
        spacing:    .shared,
        radius:     .shared,
        elevation:  .dark,
        gradients:  .shared
    )
}
